ema.c

```diff
@@ -431,87 +441,85 @@ static size_t ema_aligned_end(ema_t* ema, size_t align)
     return curr_end;
 }
 
-// Find a free space of size at least 'size' bytes, does not matter where the start is
+// Find a free space of size at least 'size' bytes on the given root, does not
+// matter where the start is
 bool find_free_region(ema_root_t *root, size_t size,
         uint64_t align, size_t *addr, ema_t **next_ema)
 {
     ema_t *ema_begin = root->guard->next;
     ema_t *ema_end = root->guard;
-    bool is_system = (root == &g_rts_ema_root);
+
+    // no ema node on the root
     if(ema_begin == ema_end){
-        size_t tmp = 0;
-        if (is_system)
-        {
-            // we need at least one node before calling this.
-            if(ema_root_empty(&g_rts_ema_root))
-                return false;//rts has to be inited at this time
-            tmp = ema_root_end(&g_rts_ema_root);
-        }else
-        {
-            tmp = mm_user_base;
+        size_t tmp = ROUND_TO(root->start, align);
+        if ((root->end - tmp) >= size) {
+            *addr = tmp;
+            *next_ema = ema_end;
+            return true;
         }
-        tmp = ROUND_TO(tmp, align);
-        if(!sgx_mm_is_within_enclave((void*)tmp, size))
-            return false;
-        *addr = tmp;
-        *next_ema = ema_end;
-        return true;
+
+        *addr = 0;
+        *next_ema = NULL;
+        return false;
     }
 
-    // iterate over the ema node within specified range
+    // iterate over the ema nodes
     ema_t *curr = ema_begin;
     ema_t *next = curr->next;
 
     while (next != ema_end) {
         size_t curr_end = ema_aligned_end(curr, align);
-        size_t free_size = next->start_addr - curr_end;
-        if (free_size >= size) {
-            *next_ema = next;
-            *addr = curr_end;
-            return true;
+        if (curr_end <= next->start_addr) {
+            size_t free_size = next->start_addr - curr_end;
+            if (free_size >= size) {
+                *next_ema = next;
+                *addr = curr_end;
+                return true;
+            }
         }
         curr = next;
         next = curr->next;
     }
 
-    // check the last ema node
-    if( sgx_mm_is_within_enclave((void*)(curr->start_addr + curr->size), size))
-    {
-        *next_ema = next;
-        *addr = ema_aligned_end(curr, align);
-        size_t end = *addr + size;
-        if( (is_system && (end <=mm_user_base || *addr > mm_user_base))
-                || (!is_system && end < mm_user_end))
+    // check the region higher than last ema node
+    size_t curr_end = ema_aligned_end(curr, align);
+    if (curr_end <= root->end) {
```

Comment:
Can we remove assumption that the RTS region is continuous? Otherwise we need introduce API to specify RTS region. Current sgx_mm_init API only specify user region. It may be beneficial to keep the flexibility of non-continuous RTS regions.